client/foundries.go

```diff
@@ -270,7 +292,54 @@ func NewApiClient(serverUrl string, config Config, caCertPath string) *Api {
 	return &api
 }
 
-func getResponse(resp *[]byte, err error, runName string) (string, string, error) {
+func httpLogger(req *http.Request) logrus.FieldLogger {
+	return logrus.WithFields(logrus.Fields{"url": req.URL.String(), "method": req.Method})
+}
+
+func readResponse(res *http.Response, log logrus.FieldLogger) (*[]byte, error) {
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		log.Debugf("I/O error reading response: %s", err)
+		return nil, err
+	}
+
+	// Accept all "normal" successful status codes: 200, 201, 202, 204, excluding quite inappropriate
+	// for RESTful web services 203, 205, and 206.  There are some preferences what to return for
+	// each operation, but a client side normally should not fail if e.g. a POST returns 200, 202, or
+	// 204 instead of a usual 201.  There are use cases when each of those status codes is valid and
+	// should be treated as a success.  Though there might be some differences how that success is
+	// handled by a higher-level logic.
+	switch res.StatusCode {
+		case 200:
+		case 201:
+		case 202:
+		case 204:
+			break
+		default:
+			var PRINT_LIMIT, DEBUG_LIMIT int = 512, 8196
+			errBody := (string)(body)
+			if len(body) > DEBUG_LIMIT {
+				// too much is too much, even for a debug message
+				errBody = fmt.Sprintf("%s...(truncated body over %d)", body[:DEBUG_LIMIT], DEBUG_LIMIT)
+			}
+			log.Debugf("HTTP error received %s", res.Status)
+			log.Debugf(errBody)
+
+			// Still return a body, a caller might need it, but also return an error
+			msg := fmt.Sprintf("HTTP error during %s '%s': %s",
+				res.Request.Method, res.Request.URL.String(), res.Status)
+			if len(body) < PRINT_LIMIT {
+				// return an error response body up to a meaningful limit - if it spans beyond a few
+				// lines, need to find a more appropriate message.
+				msg = fmt.Sprintf("%s\n=%s", msg, body)
+			}
+			err = &HttpError{msg, res}
+	}
+	return &body, err
```

Comment:
It looks like `body` is returned twice in the case of an error, as `body` and as msg in `err`, although I guess that's OK.

Reply:
Yes, in some cases it does, and that's by design.

A caller might need a body, and if it does - it should not parse it out of an error message.